services/streaming/service.go

```diff
@@ -0,0 +1,124 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package streaming
+
+import (
+	"errors"
+	"io"
+
+	api "github.com/containerd/containerd/api/services/streaming/v1"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/pkg/streaming"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/protobuf"
+	ptypes "github.com/containerd/containerd/protobuf/types"
+	"github.com/containerd/typeurl"
+	"google.golang.org/grpc"
+)
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type: plugin.GRPCPlugin,
+		ID:   "streaming",
+		Requires: []plugin.Type{
+			plugin.StreamingPlugin,
+		},
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			i, err := ic.GetByID(plugin.StreamingPlugin, "manager")
+			if err != nil {
+				return nil, err
+			}
+			return &service{manager: i.(streaming.StreamManager)}, nil
+		},
+	})
+}
+
+type service struct {
+	manager streaming.StreamManager
+	api.UnimplementedStreamingServer
+}
+
+func (s *service) Register(server *grpc.Server) error {
+	api.RegisterStreamingServer(server, s)
+	return nil
+}
+
+func (s *service) Stream(srv api.Streaming_StreamServer) error {
+	// TODO: Timeout waiting
+	a, err := srv.Recv()
+	if err != nil {
+		return err
+	}
+	var i api.StreamInit
+	if err := typeurl.UnmarshalTo(a, &i); err != nil {
+		return err
+	}
+
+	// TODO: Save this response to avoid marshaling everytime
+	response, err := typeurl.MarshalAny(&ptypes.Empty{})
+	if err != nil {
+		return err
+	}
+	if err := srv.Send(protobuf.FromAny(response)); err != nil {
```

Comment:
might be a cleaner way to do it, it could technically cause a race condition if you register first, although maybe not practically (assuming ID is always used after a response is received by the client)